core/src/main/java/org/apache/calcite/rel/rules/FilterCorrelateRule.java

```diff
@@ -86,8 +86,10 @@ public void onMatch(RelOptRuleCall call) {
         aboveFilters,
         JoinRelType.INNER,
         false,
-        !corr.getJoinType().toJoinType().generatesNullsOnLeft(),
-        !corr.getJoinType().toJoinType().generatesNullsOnRight(),
+        corr.getJoinType().returnsJustFirstInput()
+            || !corr.getJoinType().toJoinType().generatesNullsOnLeft(),
+        !corr.getJoinType().returnsJustFirstInput()
+            && !corr.getJoinType().toJoinType().generatesNullsOnRight(),
```

Comment:
I see what you are trying to do, but it really took me a while to understand. :)

Reply:
The idea is to avoid the `getJoinType().toJoinType()` call for SEMI and ANTI, which can be differentiate because they both have `getJoinType().returnsJustFirstInput() == true`